src/util/mpsc/batch.rs

```diff
@@ -0,0 +1,344 @@
+// limitations under the License.
+use std::sync::atomic::{AtomicBool, AtomicIsize, AtomicUsize, Ordering};
+pub use std::sync::mpsc::{RecvError, RecvTimeoutError, SendError, TryRecvError, TrySendError};
+use std::sync::Arc;
+use std::time::Duration;
+
+use crossbeam::sync::AtomicOption;
+use crossbeam_channel;
+use futures::task::{self, Task};
+use futures::{Async, Poll, Stream};
+
+struct State {
+    sender_cnt: AtomicIsize,
+    receiver_cnt: AtomicIsize,
+    recv_task: AtomicOption<Task>,
+    notify_size: usize,
+    // How many messages are sent without notify.
+    pending: AtomicUsize,
+    notifier_registered: AtomicBool,
+}
+
+impl State {
+    fn new(notify_size: usize) -> State {
+        State {
+            sender_cnt: AtomicIsize::new(1),
+            receiver_cnt: AtomicIsize::new(1),
+            recv_task: AtomicOption::new(),
+            notify_size,
+            pending: AtomicUsize::new(0),
+            notifier_registered: AtomicBool::new(false),
+        }
+    }
+
+    #[inline]
+    fn is_receiver_closed(&self) -> bool {
+        self.receiver_cnt.load(Ordering::Acquire) == 0
+    }
+
+    #[inline]
+    fn is_sender_closed(&self) -> bool {
+        self.sender_cnt.load(Ordering::Acquire) == 0
+    }
+
+    #[inline]
+    fn inc_send_and_maybe_notify(&self) {
+        let pending = self.pending.fetch_add(1, Ordering::SeqCst);
+        if pending >= self.notify_size - 1 {
+            self.notify();
+        }
+    }
+
+    #[inline]
+    fn notify(&self) {
+        let task = self.recv_task.take(Ordering::AcqRel);
+        if let Some(t) = task {
+            self.pending.store(0, Ordering::SeqCst);
+            t.notify();
+        }
+    }
+}
+
+/// `Notifier` is used to notify receiver whenever you want.
+pub struct Notifier(Arc<State>);
+impl Notifier {
+    #[inline]
+    pub fn notify(self) {
+        let notifier_registered = &self.0.notifier_registered;
+        if !notifier_registered.compare_and_swap(true, false, Ordering::SeqCst) {
+            unreachable!("notifier_registered must be true");
+        }
+        self.0.notify();
+    }
+}
+
+pub struct Sender<T> {
+    sender: crossbeam_channel::Sender<T>,
+    state: Arc<State>,
+}
+
+impl<T> Clone for Sender<T> {
+    #[inline]
+    fn clone(&self) -> Sender<T> {
+        self.state.sender_cnt.fetch_add(1, Ordering::AcqRel);
+        Sender {
+            sender: self.sender.clone(),
+            state: self.state.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    #[inline]
+    fn drop(&mut self) {
+        self.state.sender_cnt.fetch_add(-1, Ordering::AcqRel);
+        if self.state.is_sender_closed() {
+            self.state.notify();
+        }
+    }
+}
+
+pub struct Receiver<T> {
+    receiver: crossbeam_channel::Receiver<T>,
+    state: Arc<State>,
+}
+
+impl<T> Sender<T> {
+    #[inline]
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        if !self.state.is_receiver_closed() {
+            self.sender.send(t);
```

Comment:
How about the receiver closed before this line but after the if?

Reply:
I will use crossbeam-channel 0.3 instead of 0.2 to resolve this.